### PR TITLE
Cromwell-backend as a sub-project of Cromwell

### PR DESCRIPTION
**Pull request by francares**
1. Added Cromwell-backend library as sub-project of Cromwell.
2. Added base interfaces from Cromwell-backend.
3. Tried to use SBT Git plugin for versioning but I couldn't get it work with sbt multi-projects.

@geoffjentry and @scottfrazer could you please review this PR?

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -83,8 +83,6 @@ libraryDependencies ++= Seq(
 
 releaseSettings
 
-shellPrompt := { state => "%s| %s> ".format(GitCommand.prompt.apply(state), version.value)}
-
```
Comment (scottfrazer) [inline build.sbt]: Why remove this?

File: cromwell-backend/CHANGELOG.MD [hunk 2]
```diff
@@ -0,0 +1,5 @@
+# Cromwell-Backend Change Log
+
+1.0:
+=====
+
```
Comment (scottfrazer) [inline cromwell-backend/CHANGELOG.MD]: Are we planning on releasing cromwell-backend as a JAR by itself?  If not perhaps we should just roll this in to the top level changelog?  I must admit I'm not entirely sure what the deploy plan is for this subproject

**Comment by geoffjentry** [comment]
I'd rather see this PR purely being setting up the sub-project, not carrying over a bunch of code which will eventually live there unless everyone is 100% sure that code is exactly correct

File: cromwell-backend/CHANGELOG.MD (hunk 2 above)
Comment (geoffjentry) [inline cromwell-backend/CHANGELOG.MD]: Yes, it is a separate entity which the full cromwell jar is but one consumer of it

**Comment by scottfrazer** [comment]
Yeah I'd rather see only code committed that's actually used... can we add these classes in only when they're actually used so we're sure that everything we're adding is hooked up and tested?

**Comment by geoffjentry** [comment]
@scottfrazer thanks, that summed up my point better than my point :)

File: build.sbt (hunk 1 above)
Comment (francares) [inline build.sbt]: I removed this due to the new version of the plugin. 
I couldn't resolve the part of base version but I read that the plugin takes the base version as project.version attribute.

**Comment by francares** [comment]
@scottfrazer @geoffjentry They are going to be used in the next 3 tasks based on priority (Excel sheet), so I think It does not hurt anyone to have them as a baseline right know (they just define interfaces and not concrete implementations) since we took the work of defining them in the past.

**Comment by geoffjentry** [comment]
I'd prefer every step be just what is required for that step as we don't know exactly how things will look when subsequent steps are completed

**Comment by francares** [comment]
As you wish. I removed traits.

**Comment by francares** [comment]
@geoffjentry @scottfrazer Please, let me know if it can be merged. Thanks!

**Comment by scottfrazer** [comment]
@francares sorry for the slow responses this week.  We've got people out this week so we're falling a bit behind on code reviews (also this week was a release week, which has taken up a good amount of my time).  I will definitely review again by EOD tomorrow, but I'm not avoiding this!  I need to review @kshakir's PR first because I'm way overdue on that one, but this one is next.

**Comment by scottfrazer** [comment]
@francares So I must admit that I am not the most familiar with SBT build files.  This all seems fairly boilerplate and the parts I do know well look right to me!

My understanding of the build process now is that we still run `sbt run ...` from the top level directory and it works as expected.  On this branch, `sbt assembly` seems to create only a Cromwell JAR file.  Is this supposed to build multiple JARs or are we still making a fat-JAR with the `cromwell-backend` code part of the JAR?

I'm excited to start getting some real code in here!

:+1:

File: cromwell-backend/build.sbt [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+assemblyJarName in assembly := "cromwell-backend-" + version.value + ".jar"
+
+scalacOptions := Seq("-unchecked", "-deprecation", "-encoding", "utf8")
+
+resolvers ++= Seq("Typesafe Repository" at "http://repo.typesafe.com/typesafe/releases/",
+                  "Broad Artifactory Releases" at "https://artifactory.broadinstitute.org/artifactory/libs-release/",
+                  "Broad Artifactory Snapshots" at "https://artifactory.broadinstitute.org/artifactory/libs-snapshot/")
+
+libraryDependencies ++= {
```
Comment (geoffjentry) [inline cromwell-backend/build.sbt]: there's no code at the moment so currently no dependencies

**Comment by geoffjentry** [comment]
I only quickly looked at it but it seems to build the correct jars

**Comment by scottfrazer** [comment]
I was looking in the wrong place.  Right now it creates:

* `cromwell-backend/target/scala-2.11/cromwell-backend-0.1.jar`
* `target/scala-2.11/cromwell-0.19.jar`

And everything that's in `cromwell-backend/target/scala-2.11/cromwell-backend-0.1.jar` is included in `target/scala-2.11/cromwell-0.19.jar`

**Comment by francares** [comment]
@scottfrazer Yes, that's right. If you take a look to the build logs you will see what your are saying.

**Comment by Horneth** [comment]
:+1:  Looks good !

**Comment by francares** [comment]
@geoffjentry I removed lib dependencies in Cromwell-backend.

**Comment by scottfrazer** [comment]
Does anybody know what happens when versions of dependencies differ between `cromwell` and `cromwell-backend`?  I can find out empirically, though I wonder if there are any consequences of this we should be aware of...  and if that is a bad thing, is there any way we can define dependencies so they are in lockstep with Cromwell?

**Comment by geoffjentry** [comment]
IIRC you generally want to inherit your deps (and everything else) from the main build.sbt and only vary the things which are actually varying. 

Which raises the point that we should be doing that instead of defining all new plugins, deps, etc in here

**Comment by geoffjentry** [comment]
This is the sort of thing I'm talking about: http://stackoverflow.com/a/23189247

**Comment by scottfrazer** [comment]
Though I guess if they really are pluggable backends, then somebody can create their own SBT project and include whatever they want as long as it exposes a class that adheres to the Actor interface.  So we really can't stop differing versions of Akka or any other shared package, right?

**Comment by geoffjentry** [comment]
There's a difference between code officially supported by cromwell and stuff maintained separately. If someone else wants to wreak havoc with their own stuff that's their choice :)

**Comment by francares** [comment]
After discussion on gitter I merging it.

**Pull request merged**